Content.Client/Chat/ChatManager.cs

```diff
@@ -178,23 +179,49 @@ private void _onFilterButtonToggled(ChatBox chatBox, Button.ButtonToggledEventAr
             RepopulateChat(filteredHistory);
         }
 
-        private void RepopulateChat(List<MsgChatMessage> filteredMessages)
+        private void RepopulateChat(List<StoredChatMessage> filteredMessages)
         {
             // Copy list for enumeration
-            List<MsgChatMessage> filteredMessagesCopy = new List<MsgChatMessage>(filteredMessages);
+            List<StoredChatMessage> filteredMessagesCopy = new List<StoredChatMessage>(filteredMessages);
 
             _currentChatBox.contents.Clear();
 
-            foreach (MsgChatMessage msg in filteredMessagesCopy)
+            foreach (StoredChatMessage msg in filteredMessagesCopy)
             {
-                _onChatMessage(msg);
-                filteredMessages.Remove(msg);
+
+                if (!IsNotFiltered(msg.Channel))
+                {
+                    var color = Color.DarkGray;
+                    var messageText = msg.Message;
+                    if (!string.IsNullOrEmpty(msg.MessageWrap))
+                    {
+                        messageText = string.Format(msg.MessageWrap, messageText);
+                    }
+
+                    switch (msg.Channel)
```

Comment:
You're duplicating all this code now. What you should do is make a method that writes a `StoredChatMessage` into the chat box, and then have `_onChatMessage` use that along with storing the same chat message for later.

Reply:
I'm not sure what you mean by having `_onChatMessage` use the method for writing `StoredChatMessage` into the box. The problem is that `_onChatMessage` requires things like `MessageWrap` not present in `StoredChatMessage`. I was planning to basically make an overload for `_onChatMessage` to accept the slimmer `StoredChatMessage`.

Reply:
> The problem is that `_onChatMessage` requires things like `MessageWrap` not present in `StoredChatMessage`.

The point is to have `_onChatMessage` ONLY do what is necessary to make the  `StoredChatMessage`, and from there on the code to write it into the chat box IS identical between it and this method.

It'd roughly look like this:

```cs
private void _onChatMessage(MsgChatMessage message)
{
    Logger.Debug($"{message.Channel}: {message.Message}");

    // Log all incoming chat to repopulate when filter is untoggled
    var storedMessage = new StoredChatMessage(messageText, message.Channel);
    filteredHistory.Add(storedMessage);
    WriteChatMessage(storedMessage);
}

private void RepopulateChat(List<StoredChatMessage> filteredMessages)
{
    _currentChatBox.contents.Clear();

    foreach (var msg in filteredMessages)
    {
        WriteChatMessage(msg);
    }
}

private void WriteChatMessage(StoredChatMessage stored)
{
    if (IsNotFiltered(msg.Channel))
    {
        Logger.Debug($"Message filtered: {msg.Channel}: {msg.Message}");
        continue;
    }

    var color = Color.DarkGray;
    var messageText = msg.Message;
    if (!string.IsNullOrEmpty(msg.MessageWrap))
    {
        messageText = string.Format(msg.MessageWrap, messageText);
    }

    switch (msg.Channel)
    {
        case ChatChannel.Server:
            color = Color.Orange;
            break;
        case ChatChannel.OOC:
            color = Color.LightSkyBlue;
            break;
    }

    _currentChatBox?.AddLine(messageText, msg.Channel, color);
}
```

Reply:
Ah, so `_onChatMessage` is not called for every line you add, as they're already stored in some cases. I think I've got it now. Thanks

Reply:
Fixed!